README.md

# Platigo (Platypus Utility Golang)

Platigo is a Go utility package that provides functionality for connecting to OpenSearch, an open-source search and analytics engine. It offers a convenient way to interact with OpenSearch clusters and perform various operations such as indexing documents, creating indices, updating mappings, and executing search queries.

## Features

- **OpenSearch Integration**: Platigo enables seamless integration with OpenSearch clusters, allowing you to establish connections and interact with the search engine programmatically.
- **Indexing Documents**: You can easily index documents into the OpenSearch index using the provided utility functions.
- **Creating Indices**: Platigo offers functionality to create indices in OpenSearch, helping you manage the structure of your data.
- **Updating Mappings**: With Platigo, you can update the mapping of one or more indices in OpenSearch, ensuring proper data representation and search functionality.
- **Executing Search Queries**: Platigo allows you to perform search queries on the OpenSearch index, helping you retrieve relevant documents based on specific criteria.

## Installation

To use Platigo in your Go project, you can add it as a dependency using the following command:

```shell
go get github.com/bagastri07/platigo
```

## Usage

To get started with Platigo, create an instance of the OpenSearch client by providing the necessary configuration:

```go
config := &platigo.OSConfig{
    Addresses: []string{"http://opensearch-host:9200"},
    InsecureSkipVerify: false, // Set to true only for specific use cases where SSL certificate verification is intentionally skipped
    Username: "your-username",
    Password: "your-password",
}

client, err := platigo.NewOpensearchClient(config)
if err != nil {
    log.Fatal("Failed to create OpenSearch client:", err)
}
```

Once you have the OpenSearch client, you can use it to perform various operations. Here are a few examples:

**Indexing a Document**

```go
type MyDocument struct {
    ID      string `json:"id"`
    Title   string `json:"title"`
    Content string `json:"content"`
}

doc := &MyDocument{
    ID:      "document-id",
    Title:   "Sample Document",
    Content: "This is a sample document for Platigo.",
}

response, err := client.Index(context.Background(), "index-name", doc)
if err != nil {
    log.Fatal("Failed to index document:", err)
}
```

**Creating an Index**
```go
indexName := "new-index"

mapping := `
{
    "mappings": {
        "properties": {
            "title": {
                "type": "text"
            },
            "content": {
                "type": "text"
            }
        }
    }
}
`

response, err := client.CreateIndices(context.Background(), indexName, strings.NewReader(mapping))
if err != nil {
    log.Fatal("Failed to create index:", err)
}
```

**Performing a Search**
```go
searchQuery := `
{
    "query": {
        "match": {
            "content": "sample"
        }
    }
}
`

response, err := client.Search(context.Background(), []string{"index-name"}, strings.NewReader(searchQuery))
if err != nil {
    log.Fatal("Failed to perform search:", err)
}
```

For more details on available utility functions and their usage, please refer to the [Platigo GitHub repository](https://github.com/bagastri07/platigo).

## Contribution

Contributions to Platigo are welcome! If you encounter any issues, have suggestions for improvements, or would like to contribute new features or enhancements, please feel free to open an issue



CHANGELOG.md
# Platigo


<a name="v1.2.0"></a>
## [v1.2.0] - 2023-07-10
### New Features
- add argon encryption


<a name="v1.1.1"></a>
## [v1.1.1] - 2023-06-29
### New Features
- add ping to check cluster availability


<a name="v1.1.0"></a>
## [v1.1.0] - 2023-06-22
### New Features
- implement bulk indexer


<a name="v1.0.0"></a>
## v1.0.0 - 2023-06-16
### New Features
- add implement open search conn


[Unreleased]: https://github.com/SPS-Redigan/mantis/compare/v1.2.0...HEAD
[v1.2.0]: https://github.com/SPS-Redigan/mantis/compare/v1.1.1...v1.2.0
[v1.1.1]: https://github.com/SPS-Redigan/mantis/compare/v1.1.0...v1.1.1
[v1.1.0]: https://github.com/SPS-Redigan/mantis/compare/v1.0.0...v1.1.0


opensearch_test.go
package platigo

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestNewOpensearchClient(t *testing.T) {
	type args struct {
		config *OSConfig
	}
	tests := []struct {
		name string
		args args
	}{
		{
			name: "normal",
			args: args{
				config: &OSConfig{
					Addresses:          []string{"localhost:9200"},
					InsecureSkipVerify: true,
					Username:           "",
					Password:           "",
				},
			},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := NewOpenSearchClient(tt.args.config)
			assert.NotNil(t, got)
			assert.NoError(t, err)
		})
	}
}


crypto/argon_test.go
package crypto

import (
	"crypto/rand"
	"encoding/base64"
	"errors"
	"fmt"
	"testing"

	"github.com/agiledragon/gomonkey"
	"github.com/stretchr/testify/assert"
)

func TestHashFromPassword(t *testing.T) {
	type args struct {
		password string
	}

	tests := []struct {
		name            string
		args            args
		wantEncodedHash string
		wantErr         bool
	}{
		{
			name: "Success",
			args: args{
				password: "mysecretpassword",
			},
			wantEncodedHash: "$argon2id$v=19$m=65536,t=3,p=2$bW9ja2Vkc2FsdA$4fziyesgS8X1eccXj3E5WacLsqUeX28HHqxprFZHlY8",
			wantErr:         false,
		},
		{
			name: "Error in generateRandomBytes",
			args: args{
				password: "mysecretpassword",
			},
			wantEncodedHash: "",
			wantErr:         true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Create a monkey patch for the generateRandomBytes function.
			patch := gomonkey.ApplyFunc(generateRandomBytes, func(n uint32) ([]byte, error) {
				if tt.wantErr {
					return nil, fmt.Errorf("error in generateRandomBytes")
				}
				return []byte("mockedsalt"), nil
			})
			defer patch.Reset()

			encodedHash, err := HashFromPassword(tt.args.password)

			assert.Equal(t, tt.wantErr, err != nil, "Error status should match the expectation")
			assert.Equal(t, tt.wantEncodedHash, encodedHash, "Encoded hash should match the expected value")
		})
	}
}

func TestComparePasswordAndHash(t *testing.T) {
	type args struct {
		password string
	}

	tests := []struct {
		name            string
		args            args
		wantEncodedHash string
		wantErr         bool
		wantMatch       bool
	}{
		{
			name: "normal",
			args: args{
				password: "mysecretpassword",
			},
			wantEncodedHash: "$argon2id$v=19$m=65536,t=3,p=2$bW9ja2Vkc2FsdA$4fziyesgS8X1eccXj3E5WacLsqUeX28HHqxprFZHlY8",
			wantErr:         false,
			wantMatch:       true,
		},
		{
			name: "Error in decodeHash",
			args: args{
				password: "mysecretpassword",
			},
			wantEncodedHash: "",
			wantErr:         true,
			wantMatch:       false,
		},
		{
			name: "password not match",
			args: args{
				password: "notmatch",
			},
			wantEncodedHash: "$argon2id$v=19$m=65536,t=3,p=2$bW9ja2Vkc2FsdA$4fziyesgS8X1eccXj3E5WacLsqUeX28HHqxprFZHlY8",
			wantErr:         false,
			wantMatch:       false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			patch := gomonkey.ApplyFunc(decodeHash, func(encodedHash string) (p *params, salt, hash []byte, err error) {
				if tt.wantErr {
					return nil, nil, nil, errors.New("error in decodeHash")
				}

				// Mocked return values for successful decoding.
				p = &params{
					memory:      65536,
					iterations:  3,
					parallelism: 2,
					saltLength:  12,
					keyLength:   32,
				}
				salt, _ = base64.RawStdEncoding.Strict().DecodeString("bW9ja2Vkc2FsdA")
				hash, _ = base64.RawStdEncoding.Strict().DecodeString("4fziyesgS8X1eccXj3E5WacLsqUeX28HHqxprFZHlY8")
				return p, salt, hash, nil
			})
			defer patch.Reset()

			match, err := ComparePasswordAndHash(tt.args.password, tt.wantEncodedHash)
			assert.Equal(t, tt.wantErr, err != nil, "Error status should match the expectation")
			assert.Equal(t, tt.wantMatch, match, "Password and hash match status should match the expectation")
		})
	}
}

func TestGenerateRandomBytes(t *testing.T) {
	tests := []struct {
		name       string
		args       uint32
		wantLength int
		wantErr    bool
	}{

		{
			name:       "normal",
			args:       10,
			wantLength: 10,
			wantErr:    false,
		},
		{
			name:       "Error from rand.Read",
			args:       8,
			wantLength: 0,
			wantErr:    true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			patches := gomonkey.ApplyFunc(rand.Read, func(b []byte) (int, error) {
				if tt.wantErr {
					return 0, errors.New("err")
				}
				for i := 0; i < len(b); i++ {
					b[i] = byte(i)
				}
				return len(b), nil
			})
			defer patches.Reset()

			got, err := generateRandomBytes(tt.args)

			if tt.wantErr {
				assert.Error(t, err, "generateRandomBytes() should return an error")
				assert.Nil(t, got, "generateRandomBytes() should return a nil byte slice")
			} else {
				assert.NoError(t, err, "generateRandomBytes() should not return an error")
				assert.NotNil(t, got, "generateRandomBytes() should return a non-nil byte slice")
				assert.Equal(t, tt.wantLength, len(got), "generateRandomBytes() length mismatch")
			}
		})
	}
}

func TestDecodeHash(t *testing.T) {
	t.Run("valid hash", func(t *testing.T) {
		encodedHash := "$argon2id$v=19$m=65536,t=3,p=2$2ikdYjIFWtoN9+c87yHywA$nz2mhbQ0wqWB+ozZObSqIWBYqwlzru0mF7Ez+CFwHxE"
		expectedParams := &params{
			memory:      65536,
			iterations:  3,
			parallelism: 2,
			saltLength:  16,
			keyLength:   32,
		}
		expectedSalt, _ := base64.RawStdEncoding.Strict().DecodeString("2ikdYjIFWtoN9+c87yHywA")
		expectedHash, _ := base64.RawStdEncoding.Strict().DecodeString("nz2mhbQ0wqWB+ozZObSqIWBYqwlzru0mF7Ez+CFwHxE")

		patch := gomonkey.ApplyFunc(base64.RawStdEncoding.Strict().DecodeString, func(s string) ([]byte, error) {
			if s == "2ikdYjIFWtoN9+c87yHywA" {
				return expectedSalt, nil
			} else if s == "nz2mhbQ0wqWB+ozZObSqIWBYqwlzru0mF7Ez+CFwHxE" {
				return expectedHash, nil
			}
			return nil, errors.New("mocked error")
		})
		defer patch.Reset()

		params, salt, hash, err := decodeHash(encodedHash)
		assert.NoError(t, err)
		assert.Equal(t, expectedParams, params)
		assert.Equal(t, expectedSalt, salt)
		assert.Equal(t, expectedHash, hash)
	})

	t.Run("invalid hash - incorrect number of values", func(t *testing.T) {
		encodedHash := "dummy$dummy$dummy"
		params, salt, hash, err := decodeHash(encodedHash)
		assert.Nil(t, params)
		assert.Nil(t, salt)
		assert.Nil(t, hash)
		assert.EqualError(t, err, errInvalidHash.Error())
	})

	t.Run("invalid hash - incompatible version", func(t *testing.T) {
		encodedHash := "dummy$dummy$v=20$m=1024,t=2,p=2$dGVzdA==$dGVzdA=="
		params, salt, hash, err := decodeHash(encodedHash)
		assert.Nil(t, params)
		assert.Nil(t, salt)
		assert.Nil(t, hash)
		assert.EqualError(t, err, errIncompatibleVersion.Error())
	})

	t.Run("invalid hash - error decoding salt", func(t *testing.T) {
		encodedHash := "dummy$dummy$v=19$m=1024,t=2,p=2$dGVzdA==$dGVzdA=="
		patch := gomonkey.ApplyFunc(base64.RawStdEncoding.Strict().DecodeString, func(s string) ([]byte, error) {
			if s == "dGVzdA==" {
				return nil, errors.New("mocked error")
			}
			return nil, nil
		})
		defer patch.Reset()

		params, salt, hash, err := decodeHash(encodedHash)
		assert.Nil(t, params)
		assert.Nil(t, salt)
		assert.Nil(t, hash)
		assert.Error(t, err)
	})

	t.Run("invalid hash - error decoding hash", func(t *testing.T) {
		encodedHash := "dummy$dummy$v=19$m=1024,t=2,p=2$SGVsbG8gd29ybGQ$TWFuIGlzIGRpc3Rpbmd1aXNoZWQ="

		params, salt, hash, err := decodeHash(encodedHash)
		assert.Nil(t, params)
		assert.Nil(t, salt)
		assert.Nil(t, hash)
		assert.Error(t, err)
	})

	t.Run("invalid hash - param", func(t *testing.T) {
		encodedHash := "dummy$dummy$v=19$m=1024,t=invalid,p=2$dGVzdA==$dGVzdA=="

		params, salt, hash, err := decodeHash(encodedHash)
		assert.Nil(t, params)
		assert.Nil(t, salt)
		assert.Nil(t, hash)
		assert.Error(t, err)
	})

	t.Run("invalid hash - error get version", func(t *testing.T) {
		encodedHash := "dummy$dummy$v=invalid$m=1024,t=2,p=2$dGVzdA==$dGVzdA=="

		params, salt, hash, err := decodeHash(encodedHash)
		assert.Nil(t, params)
		assert.Nil(t, salt)
		assert.Nil(t, hash)
		assert.Error(t, err)
	})
}


opensearch.go
package platigo

import (
	"context"
	"crypto/tls"
	"net/http"
	"strings"

	"github.com/bagastri07/platigo/utils"
	"github.com/goccy/go-json"
	"github.com/opensearch-project/opensearch-go"
	"github.com/opensearch-project/opensearch-go/opensearchapi"
	"github.com/opensearch-project/opensearch-go/opensearchutil"
	"github.com/sirupsen/logrus"
)

type OSConfig struct {
	Addresses          []string
	InsecureSkipVerify bool // Set to true only if SSL certificate verification is intentionally skipped for specific use cases (e.g., testing or development).
	Username           string
	Password           string
}

type IndexModel interface {
	GetID() string
}

type OpenSearchClient interface {
	// Index indexes a document in OpenSearch.
	Index(ctx context.Context, indexName string, model IndexModel) (*opensearchapi.Response, error)

	// CreateIndices creates an index in OpenSearch.
	CreateIndices(ctx context.Context, indexName string, body *strings.Reader) (*opensearchapi.Response, error)

	// PutIndicesMapping updates the mapping for one or more indices in OpenSearch.
	PutIndicesMapping(ctx context.Context, indexNames []string, body *strings.Reader) (*opensearchapi.Response, error)

	// Search performs a search query in OpenSearch.
	Search(ctx context.Context, indexNames []string, body *strings.Reader) (*opensearchapi.Response, error)

	// BulkIndex indexes multiple documents in OpenSearch.
	BulkIndex(ctx context.Context, indexName string, models []IndexModel) error

	// Ping pings the OpenSearch cluster to check its availability.
	Ping(ctx context.Context) (*opensearchapi.Response, error)
}

type openSearchClient struct {
	client *opensearch.Client
}

// NewOpenSearchClient creates a new OpenSearchClient instance.
func NewOpenSearchClient(config *OSConfig) (OpenSearchClient, error) {
	client, err := opensearch.NewClient(opensearch.Config{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: config.InsecureSkipVerify}, // #nosec G402
		},
		Addresses: config.Addresses,
		Username:  config.Username,
		Password:  config.Password,
	})
	platigoOSClient := &openSearchClient{
		client: client,
	}

	return platigoOSClient, err
}

func (k *openSearchClient) CreateIndices(ctx context.Context, indexName string, body *strings.Reader) (*opensearchapi.Response, error) {
	logger := logrus.WithFields(logrus.Fields{
		"indexName": indexName,
	})
	req := opensearchapi.IndicesCreateRequest{
		Index: indexName,
		Body:  body,
	}

	res, err := req.Do(ctx, k.client)
	if err != nil {
		logger.Error(err.Error())
		return nil, err
	}

	logger.Info(res)

	return res, err
}

func (k *openSearchClient) PutIndicesMapping(ctx context.Context, indexNames []string, body *strings.Reader) (*opensearchapi.Response, error) {
	logger := logrus.WithFields(logrus.Fields{
		"indexNames": indexNames,
	})

	req := opensearchapi.IndicesPutMappingRequest{
		Index: indexNames,
		Body:  body,
	}

	res, err := req.Do(ctx, k.client)
	if err != nil {
		logger.Error(err.Error())
		return nil, err
	}

	logger.Info(res)

	return res, nil

}

func (k *openSearchClient) Index(ctx context.Context, indexName string, model IndexModel) (*opensearchapi.Response, error) {
	logger := logrus.WithFields(logrus.Fields{
		"indexName": indexName,
		"docID":     model.GetID(),
	})

	docData, err := json.Marshal(model)
	if err != nil {
		logger.Error(err.Error())
		return nil, err
	}

	body := strings.NewReader(string(docData))

	req := opensearchapi.IndexRequest{
		Index:      indexName,
		DocumentID: model.GetID(),
		Body:       body,
		Pretty:     true,
	}

	res, err := req.Do(ctx, k.client)
	if err != nil {
		logger.Error(err.Error())
		return nil, err
	}

	logger.Info(res)

	return res, nil
}

func (k *openSearchClient) Search(ctx context.Context, indexNames []string, body *strings.Reader) (*opensearchapi.Response, error) {
	logger := logrus.WithFields(logrus.Fields{
		"indexNames": indexNames,
	})

	req := opensearchapi.SearchRequest{
		Index:  indexNames,
		Body:   body,
		Pretty: true,
	}

	res, err := req.Do(ctx, k.client)
	if err != nil {
		logger.Error(err.Error())
		return nil, err
	}

	logger.Info(res)

	return res, nil
}

func (k *openSearchClient) BulkIndex(ctx context.Context, indexName string, models []IndexModel) error {
	logger := logrus.WithFields(logrus.Fields{
		"indexName": indexName,
	})

	bulkIndexer, err := opensearchutil.NewBulkIndexer(opensearchutil.BulkIndexerConfig{
		Index:      indexName,
		Client:     k.client,
		NumWorkers: 10,
	})

	if err != nil {
		logger.Errorf("Failed to create bulk indexer: %s", err)
		return err
	}

	for _, model := range models {
		docID := model.GetID()
		jsonData, err := json.Marshal(model)
		if err != nil {
			logger.Error(err)
			continue
		}

		item := opensearchutil.BulkIndexerItem{
			Action:     "index",
			DocumentID: docID,
			Body:       strings.NewReader(string(jsonData)),
		}

		err = bulkIndexer.Add(ctx, item)
		if err != nil {
			logger.Errorf("Failed to add document ID %s to bulk indexer: %s", docID, err)
		}
	}

	err = bulkIndexer.Close(ctx)
	if err != nil {
		logger.Errorf("Failed to close bulk indexer: %s", err)
		return err
	}

	stat := bulkIndexer.Stats()
	logger.Info("Bulk Indexer Stat: ", utils.Dump(stat))

	return nil

}

func (k *openSearchClient) Ping(ctx context.Context) (*opensearchapi.Response, error) {
	req := opensearchapi.PingRequest{}

	res, err := req.Do(ctx, k.client)
	if err != nil {
		logrus.WithError(err).Error("Failed to ping OpenSearch cluster")
		return nil, err
	}

	return res, nil
}


crypto/argon.go
package crypto

import (
	"crypto/rand"
	"crypto/subtle"
	"encoding/base64"
	"errors"
	"fmt"
	"strings"

	"golang.org/x/crypto/argon2"
)

var (
	errInvalidHash         = errors.New("the encoded hash is not in the correct format")
	errIncompatibleVersion = errors.New("incompatible version of argon2")
)

type params struct {
	memory      uint32
	iterations  uint32
	parallelism uint8
	saltLength  uint32
	keyLength   uint32
}

func HashFromPassword(password string) (encodedHash string, err error) {
	p := &params{
		memory:      64 * 1024,
		iterations:  3,
		parallelism: 2,
		saltLength:  16,
		keyLength:   32,
	}

	salt, err := generateRandomBytes(p.saltLength)
	if err != nil {
		return "", err
	}

	hash := argon2.IDKey([]byte(password), salt, p.iterations, p.memory, p.parallelism, p.keyLength)

	// Base64 encode the salt and hashed password.
	b64Salt := base64.RawStdEncoding.EncodeToString(salt)
	b64Hash := base64.RawStdEncoding.EncodeToString(hash)

	// Return a string using the standard encoded hash representation.
	encodedHash = fmt.Sprintf("$argon2id$v=%d$m=%d,t=%d,p=%d$%s$%s", argon2.Version, p.memory, p.iterations, p.parallelism, b64Salt, b64Hash)

	return encodedHash, nil
}

func generateRandomBytes(n uint32) ([]byte, error) {
	b := make([]byte, n)
	_, err := rand.Read(b)
	if err != nil {
		return nil, err
	}

	return b, nil
}

func ComparePasswordAndHash(password, encodedHash string) (match bool, err error) {
	// Extract the parameters, salt and derived key from the encoded password
	// hash.
	p, salt, hash, err := decodeHash(encodedHash)
	if err != nil {
		return false, err
	}

	// Derive the key from the other password using the same parameters.
	otherHash := argon2.IDKey([]byte(password), salt, p.iterations, p.memory, p.parallelism, p.keyLength)

	// Check that the contents of the hashed passwords are identical. Note
	// that we are using the subtle.ConstantTimeCompare() function for this
	// to help prevent timing attacks.
	if subtle.ConstantTimeCompare(hash, otherHash) == 1 {
		return true, nil
	}
	return false, nil
}

func decodeHash(encodedHash string) (p *params, salt, hash []byte, err error) {
	vals := strings.Split(encodedHash, "$")
	if len(vals) != 6 {
		return nil, nil, nil, errInvalidHash
	}

	var version int
	_, err = fmt.Sscanf(vals[2], "v=%d", &version)
	if err != nil {
		return nil, nil, nil, err
	}
	if version != argon2.Version {
		return nil, nil, nil, errIncompatibleVersion
	}

	p = &params{}
	_, err = fmt.Sscanf(vals[3], "m=%d,t=%d,p=%d", &p.memory, &p.iterations, &p.parallelism)
	if err != nil {
		return nil, nil, nil, err
	}

	salt, err = base64.RawStdEncoding.Strict().DecodeString(vals[4])
	if err != nil {
		return nil, nil, nil, err
	}
	p.saltLength = uint32(len(salt))

	hash, err = base64.RawStdEncoding.Strict().DecodeString(vals[5])
	if err != nil {
		return nil, nil, nil, err
	}
	p.keyLength = uint32(len(hash))

	return p, salt, hash, nil
}
